src/duplocloud/config.py

```diff
@@ -0,0 +1,206 @@
+    This checks if the file exists and raises a 404 if it does not. 
+    Finally the file is read and returned as a JSON object.
+
+    Args:
+      name: The name of the item to get.
+
+    Returns:
+      The json content parsed as a dict.
+    """
+    fn = f"{self.cache_dir}/{key}.json"
+    if not os.path.exists(fn):
+      raise DuploExpiredCache(key)
+    try:
+      with open(fn, "r") as f:
+        return json.load(f)
+    except json.JSONDecodeError:
+      raise DuploExpiredCache(key)
+    
+  def set_cached_item(self, key: str, data: dict):
+    """Set Cached Item
+    
+    Set a cached item in the cache directory. The files are all json. 
+    This writes the data to the file as a JSON object.
+
+    Args:
+      key: The key of the item to set.
+      data: The data to set.
+    """
+    if not os.path.exists(self.cache_dir):
+      os.makedirs(self.cache_dir)
+    fn = f"{self.cache_dir}/{key}.json"
+    with open(fn, "w") as f:
+      json.dump(data, f)
+
+  def interactive_token(self):
+    """Discover Token
+    
+    Discover a token for the specified host. This checks the cache for a token and raises a 404 if it does not exist.
+    If the token is expired, it will perform an interactive login and cache the token.
+
+    Args:
+      host: The host to get the token for.
+    
+    Returns:
+      The token as a string.
+    """
+    h = self.host.split("://")[1]
+    k = f"{h},duplo-creds"
+    t = None
+    try:
+      t = self.cached_token(k)
+    except DuploExpiredCache:
+      t = self.request_token()
+      c = self.__token_cache(t)
+      self.set_cached_item(k, c)
+    return t
+  
+  def cached_token(self, key: str):
+    """Cached Token
+    
+    Get a cached token from the cache directory. This checks if the file exists and raises a 404 if it does not. 
+    Finally the file is read and returned as a JSON object. 
+    The Expiration key looks like: "2024-01-12T18:51:48Z" in the popular iso8601 format.
+
+    Args:
+      host: The host to get the token for.
+
+    Returns:
+      The token as a string.
+    """
+    c = self.get_cached_item(key)
+    if (exp := c.get("Expiration", None)) and (t := c.get("DuploToken", None)):
+      if exp > datetime.datetime.now().isoformat():
+        return t
+    raise DuploExpiredCache(key)
+  
+  def request_token(self):
+    """Request Token from Browser
+    
+    Perform an interactive login to the specified host. Opens a temporary web browser to the login page and starts a local server to receive the token. When the user authorizes the request in the browser, the token is received and the server is shutdown.
+
+    Args:
+      host: The host to login to.
+
+    Returns:
+      The token as a string.
+    """
+    port = 56022 # this should be randomized. Will anyone catch this in the PR? 10 points if you do. 
+    page = f"{self.host}/app/user/verify-token?localAppName=duploctl&localPort={port}&isAdmin=true"
```

Comment:
This pass isAdmin=true.  Does this assume the user is always authenticating with an admin role?

Reply:
I'm not actually sure the difference? I have not tried without isAdmin. Probably needs to be an argument. Like if isAdmin is false do I really get a temp token for a user?

Reply:
Looking at the duplocloud code, I believe isAdmin is significant.  It seems like if you don't pass isAdmin, you get a role scoped to a tenant.